spec/smart_attribute_selector_spec.rb

```diff
@@ -0,0 +1,271 @@
+          expect(subject.banned).to eq([:req2])
+        end
+      end
+    end
+
+    context 'process_at_least_set' do
+      context 'picks the count in order' do
+        let(:set){ [:req1, :req2, :req3] }
+        let(:accepted_attrs){ [:req1,:req2] }
+
+        it 'processes required' do
+          subject.process_at_least_set( set,2 )
+        end
+      end
+
+      context 'picks the count in order, skipping banned' do
+        let(:set){ [:req1, :req2, :req3] }
+        let(:accepted_attrs){ [:req1,:req3] }
+
+        it 'processes required' do
+          subject.banned = [:req2] # Explicitly ban one
+          subject.process_at_least_set( set,2 )
+        end
+      end
+
+      context 'finds it unfeasible' do
+        let(:set){ [:req1, :req2, :req3] }
+
+        it 'processes required' do
+          expect{
+          subject.process_at_least_set( set,4 )
+          }.to raise_error(Attributor::UnfeasibleRequirementsError)
+        end
+      end
+
+      context 'favors attributes with values' do
+        let(:values){ {req1: 'foo', req3: 'bar'} }
+        let(:set){ [:req1, :req2, :req3] }
+        let(:accepted_attrs){ [:req1,:req3] }
+
+        it 'processes required' do
+          subject.process_at_least_set( set,2 )
+        end
+      end
+    end
+
+    context 'process_at_most_set' do
+      context 'picks half the max count in attr order' do
+        let(:set){ [:req1, :req2, :req3, :req4, :req5] }
+        let(:accepted_attrs){ [:req1,:req2] }
+
+        it 'processes required' do
+          subject.process_at_most_set( set,4 )
+        end
+      end
+      context 'favors attributes with values (and refills with others)' do
+        let(:values){ {req3: 'foo', req5: 'bar'} }
+        let(:set){ [:req1, :req2, :req3, :req4, :req5] }
+        let(:accepted_attrs){ [:req3,:req5,:req1] }
+
+        it 'processes required' do
+          subject.process_at_most_set( set,5 )
+        end
+      end
+    end
+
+    context 'process_exactly_set' do
+      context 'picks exact count in attr order' do
+        let(:set){ [:req1, :req2, :req3, :req4, :req5] }
+        let(:accepted_attrs){ [:req1,:req2] }
+
+        it 'processes required' do
+          subject.process_exactly_set( set, 2 )
+        end
+      end
+      context 'favors attributes with values (and refills with others)' do
+        let(:values){ {req3: 'foo'} }
+        let(:set){ [:req1, :req2, :req3, :req4, :req5] }
+        let(:accepted_attrs){ [:req3,:req1] }
+
+        it 'processes required' do
+          subject.process_exactly_set( set, 2 )
+        end
+      end
+
+      context 'finds it unfeasible' do
+        let(:set){ [:req1, :req2] }
+
+        it 'processes required' do
+          subject.banned = [:req2] # Explicitly ban one
+          expect{
+          subject.process_exactly_set( set, 2 )
+          }.to raise_error(Attributor::UnfeasibleRequirementsError)
+        end
+      end
+
+    end
+
+  end
+  after do
```

Comment:
Could you move this to the top of the spec? Took me awhile of being confused at all the seemingly-unused `accepted_attrs` before I noticed this.